File: FirebaseDatabase/Sources/Snapshot/FCompoundWrite.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseDatabase/Sources/Snapshot/FCompoundWrite.h"
#import "FirebaseDatabase/Sources/Core/Utilities/FImmutableTree.h"
#import "FirebaseDatabase/Sources/Core/Utilities/FPath.h"
#import "FirebaseDatabase/Sources/FNamedNode.h"
#import "FirebaseDatabase/Sources/Snapshot/FNode.h"
#import "FirebaseDatabase/Sources/Snapshot/FSnapshotUtilities.h"

@interface FCompoundWrite ()
@property(nonatomic, strong) FImmutableTree *writeTree;
@end

@implementation FCompoundWrite

- (id)initWithWriteTree:(FImmutableTree *)tree {
    self = [super init];
    if (self) {
        self.writeTree = tree;
    }
    return self;
}

+ (FCompoundWrite *)compoundWriteWithValueDictionary:
    (NSDictionary *)dictionary {
    __block FImmutableTree *writeTree = [FImmutableTree empty];
    [dictionary enumerateKeysAndObjectsUsingBlock:^(NSString *pathString,
                                                    id value, BOOL *stop) {
      id<FNode> node = [FSnapshotUtilities nodeFrom:value];
      FImmutableTree *tree = [[FImmutableTree alloc] initWithValue:node];
      writeTree = [writeTree setTree:tree
                              atPath:[[FPath alloc] initWith:pathString]];
    }];
    return [[FCompoundWrite alloc] initWithWriteTree:writeTree];
}

+ (FCompoundWrite *)compoundWriteWithNodeDictionary:(NSDictionary *)dictionary {
    __block FImmutableTree *writeTree = [FImmutableTree empty];
    [dictionary enumerateKeysAndObjectsUsingBlock:^(NSString *pathString,
                                                    id node, BOOL *stop) {
      FImmutableTree *tree = [[FImmutableTree alloc] initWithValue:node];
      writeTree = [writeTree setTree:tree
                              atPath:[[FPath alloc] initWith:pathString]];
    }];
    return [[FCompoundWrite alloc] initWithWriteTree:writeTree];
}

+ (FCompoundWrite *)emptyWrite {
    static dispatch_once_t pred = 0;
    static FCompoundWrite *empty = nil;
    dispatch_once(&pred, ^{
      empty = [[FCompoundWrite alloc]
          initWithWriteTree:[[FImmutableTree alloc] initWithValue:nil]];
    });
    return empty;
}

- (FCompoundWrite *)addWrite:(id<FNode>)node atPath:(FPath *)path {
    if (path.isEmpty) {
        return [[FCompoundWrite alloc]
            initWithWriteTree:[[FImmutableTree alloc] initWithValue:node]];
    } else {
        FTuplePathValue *rootMost =
            [self.writeTree findRootMostValueAndPath:path];
        if (rootMost != nil) {
            FPath *relativePath = [FPath relativePathFrom:rootMost.path
                                                       to:path];
            id<FNode> value = [rootMost.value updateChild:relativePath
                                             withNewChild:node];
            return [[FCompoundWrite alloc]
                initWithWriteTree:[self.writeTree setValue:value
                                                    atPath:rootMost.path]];
        } else {
            FImmutableTree *subtree =
                [[FImmutableTree alloc] initWithValue:node];
            FImmutableTree *newWriteTree = [self.writeTree setTree:subtree
                                                            atPath:path];
            return [[FCompoundWrite alloc] initWithWriteTree:newWriteTree];
        }
    }
}

- (FCompoundWrite *)addWrite:(id<FNode>)node atKey:(NSString *)key {
    return [self addWrite:node atPath:[[FPath alloc] initWith:key]];
}

- (FCompoundWrite *)addCompoundWrite:(FCompoundWrite *)compoundWrite
                              atPath:(FPath *)path {
    __block FCompoundWrite *newWrite = self;
    [compoundWrite.writeTree forEach:^(FPath *childPath, id<FNode> value) {
      newWrite = [newWrite addWrite:value atPath:[path child:childPath]];
    }];
    return newWrite;
}

/**
 * Will remove a write at the given path and deeper paths. This will
 * <em>not</em> modify a write at a higher location, which must be removed by
 * calling this method with that path.
 * @param path The path at which a write and all deeper writes should be
 * removed.
 * @return The new FWriteCompound with the removed path.
 */
- (FCompoundWrite *)removeWriteAtPath:(FPath *)path {
    if (path.isEmpty) {
        return [FCompoundWrite emptyWrite];
    } else {
        FImmutableTree *newWriteTree =
            [self.writeTree setTree:[FImmutableTree empty] atPath:path];
        return [[FCompoundWrite alloc] initWithWriteTree:newWriteTree];
    }
}

/**
 * Returns whether this FCompoundWrite will fully overwrite a node at a given
 * location and can therefore be considered "complete".
 * @param path The path to check for
 * @return Whether there is a complete write at that path.
 */
- (BOOL)hasCompleteWriteAtPath:(FPath *)path {
    return [self completeNodeAtPath:path] != nil;
}

/**
 * Returns a node for a path if and only if the node is a "complete" overwrite
 * at that path. This will not aggregate writes from depeer paths, but will
 * return child nodes from a more shallow path.
 * @param path The path to get a complete write
 * @return The node if complete at that path, or nil otherwise.
 */
- (id<FNode>)completeNodeAtPath:(FPath *)path {
    FTuplePathValue *rootMost = [self.writeTree findRootMostValueAndPath:path];
    if (rootMost != nil) {
        FPath *relativePath = [FPath relativePathFrom:rootMost.path to:path];
        return [rootMost.value getChild:relativePath];
    } else {
        return nil;
    }
}

// TODO: change into traversal method...
- (NSArray *)completeChildren {
    NSMutableArray *children = [[NSMutableArray alloc] init];
    if (self.writeTree.value != nil) {
        id<FNode> node = self.writeTree.value;
        [node enumerateChildrenUsingBlock:^(NSString *key, id<FNode> node,
                                            BOOL *stop) {
          [children addObject:[[FNamedNode alloc] initWithName:key
                                                       andNode:node]];
        }];
    } else {
        [self.writeTree.children
            enumerateKeysAndObjectsUsingBlock:^(
                NSString *childKey, FImmutableTree *childTree, BOOL *stop) {
              if (childTree.value != nil) {
                  [children addObject:[[FNamedNode alloc]
                                          initWithName:childKey
                                               andNode:childTree.value]];
              }
            }];
    }
    return children;
}

// TODO: change into enumerate method
- (NSDictionary *)childCompoundWrites {
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [self.writeTree.children
        enumerateKeysAndObjectsUsingBlock:^(
            NSString *key, FImmutableTree *childWrite, BOOL *stop) {
          dict[key] = [[FCompoundWrite alloc] initWithWriteTree:childWrite];
        }];
    return dict;
}

- (FCompoundWrite *)childCompoundWriteAtPath:(FPath *)path {
    if (path.isEmpty) {
        return self;
    } else {
        id<FNode> shadowingNode = [self completeNodeAtPath:path];
        if (shadowingNode != nil) {
            return [[FCompoundWrite alloc]
                initWithWriteTree:[[FImmutableTree alloc]
                                      initWithValue:shadowingNode]];
        } else {
            return [[FCompoundWrite alloc]
                initWithWriteTree:[self.writeTree subtreeAtPath:path]];
        }
    }
}

- (id<FNode>)applySubtreeWrite:(FImmutableTree *)subtreeWrite
                        atPath:(FPath *)relativePath
                        toNode:(id<FNode>)node {
    if (subtreeWrite.value != nil) {
        // Since a write there is always a leaf, we're done here.
        return [node updateChild:relativePath withNewChild:subtreeWrite.value];
    } else {
        __block id<FNode> priorityWrite = nil;
        __block id<FNode> blockNode = node;
        [subtreeWrite.children
            enumerateKeysAndObjectsUsingBlock:^(
                NSString *childKey, FImmutableTree *childTree, BOOL *stop) {
              if ([childKey isEqualToString:@".priority"]) {
                  // Apply priorities at the end so we don't update priorities
                  // for either empty nodes or forget to apply priorities to
                  // empty nodes that are later filled.
                  NSAssert(childTree.value != nil,
                           @"Priority writes must always be leaf nodes");
                  priorityWrite = childTree.value;
              } else {
                  blockNode = [self
                      applySubtreeWrite:childTree
                                 atPath:[relativePath childFromString:childKey]
                                 toNode:blockNode];
              }
            }];
        // If there was a priority write, we only apply it if the node is not
        // empty
        if (![blockNode getChild:relativePath].isEmpty &&
            priorityWrite != nil) {
            blockNode = [blockNode
                 updateChild:[relativePath childFromString:@".priority"]
                withNewChild:priorityWrite];
        }
        return blockNode;
    }
}

- (void)enumerateWrites:(void (^)(FPath *, id<FNode>, BOOL *))block {
    __block BOOL stop = NO;
    // TODO: add stop to tree iterator...
    [self.writeTree forEach:^(FPath *path, id value) {
      if (!stop) {
          block(path, value, &stop);
      }
    }];
}

/**
 * Applies this FCompoundWrite to a node. The node is returned with all writes
 * from this FCompoundWrite applied to the node.
 * @param node The node to apply this FCompoundWrite to
 * @return The node with all writes applied
 */
- (id<FNode>)applyToNode:(id<FNode>)node {
    return [self applySubtreeWrite:self.writeTree
                            atPath:[FPath empty]
                            toNode:node];
}

/**
 * Return true if this CompoundWrite is empty and therefore does not modify any
 * nodes.
 * @return Whether this CompoundWrite is empty
 */
- (BOOL)isEmpty {
    return self.writeTree.isEmpty;
}

- (id<FNode>)rootWrite {
    return self.writeTree.value;
}

- (BOOL)isEqual:(id)object {
    if (![object isKindOfClass:[FCompoundWrite class]]) {
        return NO;
    }
    FCompoundWrite *other = (FCompoundWrite *)object;
    return
        [[self valForExport:YES] isEqualToDictionary:[other valForExport:YES]];
}

- (NSUInteger)hash {
    return [[self valForExport:YES] hash];
}

- (NSDictionary *)valForExport:(BOOL)exportFormat {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [self.writeTree forEach:^(FPath *path, id<FNode> value) {
      dictionary[path.wireFormat] = [value valForExport:exportFormat];
    }];
    return dictionary;
}

- (NSString *)description {
    return [[self valForExport:YES] description];
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FCompoundWrite.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 28/09/2021.
//

import Foundation
/**
 * This class holds a collection of writes that can be applied to nodes in
 * unison. It abstracts away the logic with dealing with priority writes and
 * multiple nested writes. At any given path, there is only allowed to be one
 * write modifying that path. Any write to an existing path or shadowing an
 * existing path will modify that existing write to reflect the write added.
 */
@objc public class FCompoundWrite: NSObject {
    let writeTree: FImmutableTreeSwift<FNode>
    init(writeTree: FImmutableTreeSwift<FNode>) {
        self.writeTree = writeTree
    }

    /**
     * Creates a compound write with NSDictionary from path string to object
     */
    @objc public static func compoundWrite(valueDictionary dictionary: NSDictionary) -> FCompoundWrite {
        var writeTree: FImmutableTreeSwift<FNode> = .empty()
        dictionary.enumerateKeysAndObjects { pathString, value, _ in
            guard let path = pathString as? String else { return }
            let node = FSnapshotUtilitiesSwift.nodeFrom(value)
            let tree = FImmutableTreeSwift<FNode>(value: node)
            writeTree = writeTree.setTree(tree, atPath: FPath(with: path))
        }
        return FCompoundWrite(writeTree: writeTree)
    }

    @objc public static func compoundWrite(nodeDictionary dictionary: NSDictionary) -> FCompoundWrite {
        var writeTree = FImmutableTreeSwift<FNode>.empty()
        dictionary.enumerateKeysAndObjects { key, value, _ in
            guard let pathString = key as? String else { return }
            guard let node = value as? FNode else { return }
            let tree = FImmutableTreeSwift(value: node)
            writeTree = writeTree.setTree(tree, atPath: FPath(with: pathString))
        }
        return FCompoundWrite(writeTree: writeTree)
    }

    @objc public static let emptyWrite: FCompoundWrite = FCompoundWrite(writeTree: .empty())

    @objc public func addWrite(_ node: FNode, atPath path: FPath) -> FCompoundWrite {
        if path.isEmpty() {
            return FCompoundWrite(writeTree: FImmutableTreeSwift(value: node))
        } else {
            if let rootMost = writeTree.findRootMostValueAndPath(path) {
                let relativePath = FPath.relativePath(from: rootMost.path, to: path)
                let value = rootMost.value.updateChild(relativePath, withNewChild: node)
                return FCompoundWrite(writeTree: self.writeTree.setValue(value, atPath: rootMost.path))
            } else {
                let subtree = FImmutableTreeSwift<FNode>(value: node)
                let newWriteTree = self.writeTree.setTree(subtree, atPath: path)
                return FCompoundWrite(writeTree: newWriteTree)
            }
        }
    }

    @objc public func addWrite(_ node: FNode, atKey key: String) -> FCompoundWrite {
        addWrite(node, atPath: FPath(with: key))
    }

    @objc public func addCompoundWrite(_ compoundWrite: FCompoundWrite, atPath path: FPath) -> FCompoundWrite {
        var newWrite = self
        compoundWrite.writeTree.forEach { childPath, value in
            newWrite = newWrite.addWrite(value, atPath: path.child(childPath))
        }
        return newWrite
    }

    /**
     * Will remove a write at the given path and deeper paths. This will
     * <em>not</em> modify a write at a higher location, which must be removed by
     * calling this method with that path.
     * @param path The path at which a write and all deeper writes should be
     * removed.
     * @return The new FWriteCompound with the removed path.
     */
    @objc public func removeWriteAtPath(_ path: FPath) -> FCompoundWrite {
        if path.isEmpty() {
            return FCompoundWrite.emptyWrite
        } else {
            let newWriteTree = self.writeTree.setTree(.empty(), atPath: path)
            return FCompoundWrite(writeTree: newWriteTree)
        }
    }

    @objc public var rootWrite: FNode? {
        writeTree.value
    }

    /**
     * Returns whether this FCompoundWrite will fully overwrite a node at a given
     * location and can therefore be considered "complete".
     * @param path The path to check for
     * @return Whether there is a complete write at that path.
     */
    @objc public func hasCompleteWriteAtPath(_ path: FPath) -> Bool {
        completeNodeAtPath(path) != nil
    }

    /**
     * Returns a node for a path if and only if the node is a "complete" overwrite
     * at that path. This will not aggregate writes from depeer paths, but will
     * return child nodes from a more shallow path.
     * @param path The path to get a complete write
     * @return The node if complete at that path, or nil otherwise.
     */
    @objc public func completeNodeAtPath(_ path: FPath) -> FNode? {
        guard let rootMost = self.writeTree.findRootMostValueAndPath(path) else {
            return nil
        }
        let relativePath = FPath.relativePath(from: rootMost.path, to: path)
        return rootMost.value.getChild(relativePath)
    }

    // TODO: change into traversal method...
    @objc public var completeChildren: [FNamedNode] {
        var children: [FNamedNode] = []
        if let node = writeTree.value {
            node.enumerateChildren { key, node, _ in
                children.append(FNamedNode(name: key, andNode: node))
            }
        } else {
            writeTree.forEachChild { childKey, childValue in
                if let value = childValue {
                    children.append(FNamedNode(name: childKey, andNode: value))
                }
            }
        }
        return children
    }

    @objc public var childCompoundWrites: [String: FCompoundWrite] {
        var dict: [String: FCompoundWrite] = [:]
        writeTree.forEachChildTree { childKey, childTree in
            dict[childKey] = FCompoundWrite(writeTree: childTree)
        }
        return dict
    }

    @objc public func childCompoundWriteAtPath(_ path: FPath) -> FCompoundWrite {
        if path.isEmpty() {
            return self
        } else {
            if let shadowingNode = self.completeNodeAtPath(path) {
                return FCompoundWrite(writeTree: FImmutableTreeSwift(value: shadowingNode))
            } else {
                return FCompoundWrite(writeTree: writeTree.subtree(atPath: path))
            }
        }
    }

    func applySubtreeWrite(_ subtreeWrite: FImmutableTreeSwift<FNode>, atPath relativePath: FPath, toNode node: FNode) -> FNode {
        if let value = subtreeWrite.value {
            // Since a write there is always a leaf, we're done here.
            return node.updateChild(relativePath, withNewChild: value)
        } else {
            var priorityWrite: FNode? = nil
            var blockNode: FNode = node
            subtreeWrite.forEachChildTree { childKey, childTree in
                if childKey == ".priority" {
                    // Apply priorities at the end so we don't update priorities
                    // for either empty nodes or forget to apply priorities to
                    // empty nodes that are later filled.
                    assert(childTree.value != nil,
                             "Priority writes must always be leaf nodes")
                    priorityWrite = childTree.value

                } else {
                    blockNode = self.applySubtreeWrite(childTree, atPath: relativePath.child(fromString: childKey), toNode: blockNode)
                }
            }
            // If there was a priority write, we only apply it if the node is not
            // empty
            if let priorityWrite = priorityWrite, !blockNode.getChild(relativePath).isEmpty {
                blockNode = blockNode.updateChild(relativePath.child(fromString: ".priority"),
                                                  withNewChild:priorityWrite)
            }
            return blockNode
        }
    }

    /**
     * Applies this FCompoundWrite to a node. The node is returned with all writes
     * from this FCompoundWrite applied to the node.
     * @param node The node to apply this FCompoundWrite to
     * @return The node with all writes applied
     */
    @objc public func applyToNode(_ node: FNode) -> FNode {
        applySubtreeWrite(self.writeTree,
                          atPath:FPath.empty(),
                          toNode:node)

    }

    @objc public func enumerateWrites(_ block: @escaping (FPath, FNode,UnsafeMutablePointer<ObjCBool>) -> Void) {
        var stop: ObjCBool = false
        // TODO: add stop to tree iterator...
        writeTree.forEach { path, value in
            if !stop.boolValue {
                block(path, value, &stop)
            }
        }
    }

    @objc public func valForExport(_ exportFormat: Bool) -> NSDictionary {
        let dictionary = NSMutableDictionary()
        writeTree.forEach { path, value in
            dictionary[path.wireFormat()] = value.val(forExport: exportFormat)
        }
        return dictionary
    }

    /**
     * Return true if this CompoundWrite is empty and therefore does not modify any
     * nodes.
     * @return Whether this CompoundWrite is empty
     */
    @objc public var isEmpty: Bool {
        writeTree.isEmpty()
    }

    public override var description: String {
        valForExport(true).description
    }

    public override func isEqual(_ object: Any?) -> Bool {
        guard let other = object as? FCompoundWrite else { return false }
        return valForExport(true).isEqual(other.valForExport(true))
    }

    public override var hash: Int {
        valForExport(true).hash
    }
}
